bugsnag-android-performance/src/main/kotlin/com/bugsnag/android/performance/internal/framerate/FramerateCollector.kt

```diff
@@ -0,0 +1,142 @@
+package com.bugsnag.android.performance.internal.framerate
+
+import android.os.Build
+import android.os.SystemClock
+import android.view.Choreographer
+import android.view.FrameMetrics
+import android.view.Window
+import androidx.annotation.RequiresApi
+
+@RequiresApi(Build.VERSION_CODES.N)
+internal abstract class FramerateCollector(
+    private val metricsContainer: FramerateMetricsContainer,
+) : Window.OnFrameMetricsAvailableListener {
+
+    private var previousFrameTime: Long = 0L
+
+    override fun onFrameMetricsAvailable(
+        window: Window,
+        frameMetrics: FrameMetrics,
+        dropCountSinceLastInvocation: Int,
+    ) {
+        val frameStartTime = frameMetrics.frameStart
+        // early-exit if this appears to be a duplicate frame, or the first frame
+        if (frameStartTime < previousFrameTime) {
+            return
+        } else if (frameMetrics.isFirstFrame) {
+            previousFrameTime = frameStartTime
+            return
+        }
+
+        // count the number of times onFrameMetricsAvailable is called - this *must* be done first
+        metricsContainer.totalMetricsCount++
+
+        val totalDuration = frameMetrics.totalDuration
+        val deadline = frameMetrics.deadline
+
+        // we allow for a 5% overrun when considering "slow" frames
+        val adjustedDeadline: Long = (deadline * SLOW_FRAME_ADJUSTMENT).toLong()
+
+        if (totalDuration >= FROZEN_FRAME_TIME) {
+            val spanStartTime = frameTimestampToClockTime(frameStartTime)
+            metricsContainer.addFrozenFrame(spanStartTime - totalDuration, spanStartTime)
```

Comment:
Slightly academic point, but it's the end time of the span I believe?

```suggestion
            val spanEndTime = frameTimestampToClockTime(frameStartTime)
            metricsContainer.addFrozenFrame(spanEndTime - totalDuration, spanEndTime)
```